src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -55,20 +57,29 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
         private readonly Lazy<IVsPackageUninstaller> _packageUninstaller;
         private readonly Lazy<IVsPackageSourceProvider> _packageSourceProvider;
 
-        private JoinableTask<ImmutableArray<PackageSource>?> _packageSourcesAsync;
         private IVsPackage _nugetPackageManager;
 
-        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
-
         // We keep track of what types of changes we've seen so we can then determine what to
         // refresh on the UI thread.  If we hear about project changes, we only refresh that
         // project.  If we hear about a solution level change, we'll refresh all projects.
-        private bool _solutionChanged;
-        private readonly HashSet<ProjectId> _changedProjects = new HashSet<ProjectId>();
+        private AsyncBatchingWorkQueue<(bool solutionChanged, ProjectId changedProject)> _workQueue;
 
         private readonly ConcurrentDictionary<ProjectId, ProjectState> _projectToInstalledPackageAndVersion =
             new ConcurrentDictionary<ProjectId, ProjectState>();
 
+        /// <summary>
+        /// Lock used to protect reads and writes of <see cref="_packageSourcesTask"/>.
+        /// </summary>
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// Task uses to compute the set of package sources on demand when asked the first time.  The value will be
+        /// computed and cached in the task.  When this value changes, the task will simply be cleared out.
+        /// </summary>
+        private Task<ImmutableArray<PackageSource>> _packageSourcesTask;
+
+        private Shell.IAsyncServiceProvider AsyncServiceProvider => (Shell.IAsyncServiceProvider)_serviceProvider;
```

Comment:
I don't think this cast (is guaranteed to) work. You have to query the `IServiceProvider` for an `SAsyncServiceProvider` service. The reason being that the former is an STA object while the latter is a free-threaded object. So they are actually different objects so the one must be queried from the other.

Reply:
> I don't think this cast (is guaranteed to) work. You have to query the IServiceProvider for an SAsyncServiceProvider service

Interesting.  This is waht we do elsewhere in roslyn, so that now seems suspect.

Reply:
It really depends on where _serviceProvider comes from, but managed code rarely/never has access to the raw STA service provider (they can get it, but it isn't very useful since it doesn't offer any more than they have with their managed IServiceProvider impl) in managed code it is almost always either:

1) An instance of the managed ServiceProvider/AsyncServiceProvider type from MPF

2) An instance of the Package/AsyncPackage type from MPF

For ServiceProvider/Package all they implement is the sync IServiceProvider interface, so the cast against either of those would fail. AsyncServiceProvider doesn't implement IServiceProvider so if _serviceProvider is typed as IServiceProvider it couldn't even be stored in that field (I didn't check the backing type, just going by this small code snippet out of context). AsyncPackage implements both IServiceProvider (via its derivation from Package) and IAsyncServiceProvider, so this cast is valid.

That said, 'horizontal casting' like this, i.e. casting between otherwise unrelated types, smells of implementation detail leakage. It's one thing to assume anyone that impements IFoo2 also implements IFoo, but assuming anyone that implements IFoo also implements IBar seems different.

I think the most 'pure' way is as @AArnott suggests, explicitly querying for the service you need from the known thing you have, i.e. asking for SAsyncServiceProvider from IServiceProvider.

Reply:
Another question for Ryan Molden then, I guess. I'm not sure. I just thought I remembered the cast failing and having to query for it. Maybe I didn't have the *global* `IServiceProvider` at the time. Certainly if you had some off-shoot (e.g. an MPF Package-level ISP instance) then casting would fail. So in general, querying for it seems safer unless you know exactly which instance of ISP you have.

Reply:
It is probably because they pass around their package instance as the IServiceProvider and it is an AsyncPackage derived package, in that case it implements both IServiceProvider and IAsyncServiceProvider, and delegates unrecognized service requests to the global provider (after siting/before unsiting). But it implementing both is more of an implementation detail of how service proferring works in VS (i.e. you register an IServiceProvider/IAsyncServiceProvider instance as a callback point to retrieve the services you own) than any sort of formal contract around interface 'relatedness'